File: day3.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.regex.Pattern;

public class Main {
    public static void main(String[] args) throws FileNotFoundException {
        System.out.println(processInput(getInput(), true));
        System.out.println(processInput(getInput(), false));
    }
    private static int processInput(char[][] input, boolean part1) {
        var sum = 0;
        for(var i=0; i<input.length; i++){//traverse through each line
            for(var j=0; j<input[0].length; j++){
                if(part1 && !Pattern.matches("[0-9.]", String.valueOf(input[i][j]))){
                    sum += calculatePieces(input, i, j, true);
                }
                else if(!part1 && Pattern.matches("\\*", String.valueOf(input[i][j]))){
                    sum += calculatePieces(input, i, j, false);
                }
            }
        }
        return sum;
    }
        private static int calculatePieces(char[][] input, int row, int col, boolean part1) {
        var sum = part1 ? 0 : 1;
        var cnt = 0;
        int[] dx = {1, -1, 0, 0, 1, -1, 1, -1};
        int[] dy = {0, 0, 1, -1, 1, -1, -1, 1};

        for (var d = 0; d < 8; d++) {
            var newRow = row + dx[d];
            var newCol = col + dy[d];

            if (newRow >= 0 && newRow < input.length && newCol >= 0 && newCol < input[newRow].length) {
                var adjacentChar = input[newRow][newCol];
                var orCol = newCol;
                if (Character.isDigit(adjacentChar)) {
                    StringBuilder num = new StringBuilder(String.valueOf(adjacentChar));
                    input[newRow][newCol] = '.';
                    while(newCol + 1 < input.length && Character.isDigit(input[newRow][newCol+1])){
                        num.append(input[newRow][newCol + 1]);
                        input[newRow][newCol+1] = '.';
                        newCol++;
                    }
                    newCol = orCol;
                    while(newCol - 1 >= 0 && Character.isDigit(input[newRow][newCol-1])){
                        num.insert(0, input[newRow][newCol - 1]);
                        input[newRow][newCol-1] = '.';
                        newCol--;
                    }
                    if(part1)
                        sum+= Integer.parseInt(num.toString());
                    else {
                        sum *= Integer.parseInt(num.toString());
                        cnt++;
                    }
                }
            }
        }
            if(!part1 && cnt == 1)
                return 0;
        return sum;
    }




    public static char[][] getInput() throws FileNotFoundException {
        var reader = new Scanner(new File("C:\\Users\\USER\\IdeaProjects\\Advent2023\\src\\input3.txt"));
        var list = new ArrayList<String>();
        while (reader.hasNextLine())
            list.add(reader.nextLine());

        char[][] input = new char[list.size()][list.get(0).length()];

        for (var i = 0; i < list.size(); i++) {
            char[] chars = list.get(i).toCharArray();
            System.arraycopy(chars, 0, input[i], 0, chars.length);
        }

        return input;
    }
}

File: day3.py
import re


def process_input(input_data, part1):
    rows, cols = len(input_data), len(input_data[0])
    total_sum = 0

    for i in range(rows):
        for j in range(cols):
            if part1 and not re.match("[0-9.]", input_data[i][j]):
                total_sum += calculate_pieces(input_data, i, j, True)
            elif not part1 and re.match("\\*", input_data[i][j]):
                total_sum += calculate_pieces(input_data, i, j, False)

    return total_sum


def calculate_pieces(input_data, row, col, part1):
    total_sum = 0 if part1 else 1
    count = 0
    dx, dy = [1, -1, 0, 0, 1, -1, 1, -1], [0, 0, 1, -1, 1, -1, -1, 1]

    for x, y in zip(dx, dy):
        new_row, new_col = row + x, col + y

        adjacent_char = input_data[new_row][new_col]
        original_col = new_col

        if adjacent_char.isdigit():
            num = [adjacent_char]
            input_data[new_row][new_col] = '.'

            while new_col + 1 < len(input_data) and input_data[new_row][new_col + 1].isdigit():
                num.append(input_data[new_row][new_col + 1])
                input_data[new_row][new_col + 1] = '.'
                new_col += 1

            new_col = original_col

            while new_col - 1 >= 0 and input_data[new_row][new_col - 1].isdigit():
                num.insert(0, input_data[new_row][new_col - 1])
                input_data[new_row][new_col - 1] = '.'
                new_col -= 1

            if part1:
                total_sum += int(''.join(num))
            else:
                total_sum *= int(''.join(num))
                count += 1

    if not part1 and count == 1:
        return 0

    return total_sum


def get_input():
    with open("input3.txt", "r") as file:
        return [list(line.strip()) for line in file]


print(process_input(get_input(), True))
print(process_input(get_input(), False))
